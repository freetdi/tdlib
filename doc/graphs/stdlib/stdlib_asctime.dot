digraph G {
0[label="0, 0: "];
1[label="1, 1: "];
2[label="2, 2: "];
3[label="3, 3: "];
4[label="4, 4: 0 1 "];
5[label="5, 5: 0 1 2 3 "];
6[label="6, 6: 0 1 2 3 4 5 "];
7[label="7, 7: 0 1 4 5 6 7 "];
8[label="8, 10: 0 1 6 7 8 "];
9[label="9, 11: 0 1 6 7 8 9 10 "];
10[label="10, 13: 0 1 6 7 9 10 11 12 "];
11[label="11, 14: 0 1 6 7 9 10 11 12 13 "];
12[label="12, 15: 0 1 6 7 9 10 13 14 15 "];
13[label="13, 17: 0 1 6 7 9 10 14 15 16 17 "];
14[label="14, 18: 0 1 6 7 9 10 14 15 16 17 18 "];
15[label="15, 19: 0 1 6 7 9 10 14 15 18 19 20 "];
16[label="16, 21: 0 1 6 7 9 10 14 15 19 20 21 22 "];
17[label="17, 22: 0 1 6 7 9 10 14 15 19 20 21 22 23 "];
18[label="18, 23: 0 1 6 7 9 10 14 15 19 20 23 24 25 "];
19[label="19, 24: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 "];
20[label="20, 26: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 28 29 "];
21[label="21, 27: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 28 29 30 "];
22[label="22, 28: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 30 31 32 "];
23[label="23, 29: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 31 32 33 34 "];
24[label="24, 30: 0 1 6 7 9 10 14 15 19 20 24 25 26 27 33 34 35 36 "];
25[label="25, 31: 0 1 6 7 9 10 14 15 19 20 24 25 35 36 37 38 "];
26[label="26, 32: 0 1 6 7 9 10 14 15 19 20 24 25 37 38 39 40 "];
27[label="27, 34: 0 1 6 7 9 10 14 15 19 20 24 25 37 38 39 40 41 42 "];
28[label="28, 35: 6 7 9 10 14 15 19 20 24 25 37 38 39 40 41 42 43 "];
29[label="29, 36: 6 7 9 10 14 15 19 20 24 25 37 38 39 40 43 44 45 "];
30[label="30, 37: 6 7 9 10 14 15 19 20 24 25 37 38 39 40 44 45 46 47 "];
31[label="31, 38: 6 7 9 10 14 15 19 20 24 25 37 38 39 40 46 47 48 49 "];
32[label="32, 39: 6 7 9 10 14 15 19 20 24 25 37 38 48 49 50 51 "];
33[label="33, 40: 6 7 9 10 14 15 19 20 24 25 37 38 50 51 52 53 "];
34[label="34, 41: 6 7 9 10 14 15 19 20 24 25 37 38 50 51 52 53 54 55 "];
35[label="35, 42: 6 7 9 10 14 15 19 20 24 25 37 38 50 51 54 55 56 57 "];
36[label="36, 43: 6 7 9 10 14 15 19 20 24 25 37 38 50 51 54 55 56 57 58 59 "];
37[label="37, 44: 6 7 9 10 14 15 19 20 24 25 37 38 50 51 54 55 56 57 58 59 "];
38[label="38, 45: 9 10 14 15 19 20 24 25 37 38 50 51 54 55 56 57 58 59 "];
39[label="39, 46: 14 15 19 20 24 25 37 38 50 51 54 55 56 57 58 59 "];
40[label="40, 47: 19 20 24 25 37 38 50 51 54 55 56 57 58 59 "];
41[label="41, 48: 24 25 37 38 50 51 54 55 56 57 58 59 "];
42[label="42, 49: 37 38 50 51 54 55 56 57 58 59 "];
43[label="43, 50: 50 51 54 55 56 57 58 59 "];
44[label="44, 51: 54 55 56 57 58 59 "];
45[label="45, 52: 56 57 58 59 "];
46[label="46, 53: 56 57 "];
47[label="47, 54: 56 57 60 61 "];
48[label="48, 55: 60 61 62 63 "];
49[label="49, 56: 62 63 "];
50[label="50, 57: "];
51[label="51, 58: "];
0->1 ;
1->2 ;
2->3 ;
3->4 ;
4->5 ;
5->6 ;
6->7 ;
7->8 ;
8->9 ;
9->10 ;
10->11 ;
11->12 ;
12->13 ;
13->14 ;
14->15 ;
15->16 ;
16->17 ;
17->18 ;
18->19 ;
19->20 ;
20->21 ;
21->22 ;
22->23 ;
23->24 ;
24->25 ;
25->26 ;
26->27 ;
27->28 ;
28->29 ;
29->30 ;
30->31 ;
31->32 ;
32->33 ;
33->34 ;
34->35 ;
35->36 ;
36->37 ;
37->38 ;
38->39 ;
39->40 ;
40->41 ;
41->42 ;
42->43 ;
43->44 ;
44->45 ;
45->46 ;
46->47 ;
47->48 ;
48->49 ;
49->50 ;
50->51 ;
}
