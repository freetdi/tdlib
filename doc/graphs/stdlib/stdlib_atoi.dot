digraph G {
0[label="0, 0: "];
1[label="1, 1: "];
2[label="2, 2: 0 1 "];
3[label="3, 105: 0 1 2 3 "];
4[label="4, 4: 0 1 2 3 "];
5[label="5, 111: 0 1 2 3 4 "];
6[label="6, 112: 0 1 2 3 4 5 "];
7[label="7, 113: 0 1 2 3 4 5 6 "];
8[label="8, 7: 0 1 2 3 4 5 6 "];
9[label="9, 12: 0 1 2 3 4 5 6 "];
10[label="10, 13: 0 1 2 3 4 5 6 "];
11[label="11, 18: 0 1 2 3 4 5 6 "];
12[label="12, 19: 0 1 2 3 5 6 "];
13[label="13, 23: 0 1 2 3 5 6 "];
14[label="14, 27: 0 1 2 3 5 6 "];
15[label="15, 33: 0 1 2 3 5 6 "];
16[label="16, 39: 0 1 2 3 "];
17[label="17, 41: 0 1 2 3 "];
18[label="18, 106: 0 1 2 3 5 6 "];
19[label="19, 107: 0 1 2 3 5 6 "];
20[label="20, 43: 0 1 2 3 5 6 7 8 "];
21[label="21, 46: 0 1 5 6 7 8 9 "];
22[label="22, 50: 0 1 5 6 7 8 9 "];
23[label="23, 56: 0 1 6 7 8 9 "];
24[label="24, 58: 0 1 7 8 9 "];
25[label="25, 60: 0 1 7 8 9 10 11 "];
26[label="26, 61: 0 1 9 10 11 "];
27[label="27, 104: 0 1 9 "];
28[label="28, 108: 0 1 9 12 13 "];
29[label="29, 63: 0 1 9 12 13 "];
30[label="30, 65: 0 1 9 12 13 14 "];
31[label="31, 66: 0 1 9 12 13 14 "];
32[label="32, 71: 0 1 9 12 13 14 "];
33[label="33, 72: 0 1 9 12 13 14 "];
34[label="34, 77: 0 1 9 12 13 14 "];
35[label="35, 78: 0 1 9 12 13 14 "];
36[label="36, 120: 0 1 9 12 13 14 "];
37[label="37, 121: 9 12 13 14 "];
38[label="38, 122: 9 12 13 14 15 16 "];
39[label="39, 84: 9 12 13 14 15 16 17 18 "];
40[label="40, 85: 9 12 13 15 16 17 18 19 20 "];
41[label="41, 86: 9 12 13 15 16 19 20 21 22 "];
42[label="42, 87: 0 1 9 12 13 21 22 "];
43[label="43, 89: 0 1 9 12 13 "];
44[label="44, 91: 0 1 9 12 13 "];
45[label="45, 92: 0 1 9 "];
46[label="46, 93: 0 1 9 "];
47[label="47, 94: 0 1 23 24 "];
48[label="48, 95: 23 24 25 26 "];
49[label="49, 96: 25 26 "];
50[label="50, 97: 0 1 "];
51[label="51, 98: 0 1 25 26 "];
52[label="52, 99: 25 26 "];
53[label="53, 100: 25 26 "];
54[label="54, 101: "];
55[label="55, 102: "];
0->1 ;
1->2 ;
2->3 ;
3->4 ;
4->5 ;
5->6 ;
6->7 ;
7->8 ;
8->9 ;
8->18 ;
9->10 ;
10->11 ;
10->13 ;
11->12 ;
12->13 ;
12->18 ;
13->14 ;
14->15 ;
14->18 ;
15->16 ;
15->18 ;
16->17 ;
17->4 ;
18->19 ;
19->20 ;
20->21 ;
21->22 ;
22->23 ;
22->24 ;
23->24 ;
23->27 ;
24->25 ;
25->26 ;
26->27 ;
27->28 ;
28->29 ;
29->30 ;
30->31 ;
31->32 ;
31->45 ;
32->33 ;
33->34 ;
33->45 ;
34->35 ;
35->36 ;
35->45 ;
36->37 ;
37->38 ;
38->39 ;
39->40 ;
40->41 ;
41->42 ;
42->43 ;
43->44 ;
44->29 ;
45->46 ;
46->47 ;
46->50 ;
47->48 ;
48->49 ;
49->52 ;
50->51 ;
51->52 ;
52->53 ;
53->54 ;
54->55 ;
}
