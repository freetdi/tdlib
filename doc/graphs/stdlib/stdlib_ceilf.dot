digraph G {
0[label="0, 0: "];
1[label="1, 1: "];
2[label="2, 29: "];
3[label="3, 30: 0 1 2 3 "];
4[label="4, 3: 0 1 2 3 4 5 6 7 "];
5[label="5, 31: 4 5 6 7 "];
6[label="6, 32: 4 5 6 7 8 9 10 11 "];
7[label="7, 4: 4 5 6 7 8 9 10 11 "];
8[label="8, 5: 4 5 6 7 8 9 10 11 "];
9[label="9, 9: 8 9 10 11 "];
10[label="10, 11: 4 5 6 7 8 9 10 11 "];
11[label="11, 33: 4 5 6 7 8 9 10 11 "];
12[label="12, 34: 4 5 6 7 8 9 10 11 "];
13[label="13, 35: 4 5 6 7 12 "];
14[label="14, 14: 4 5 6 7 12 "];
15[label="15, 15: 4 5 6 7 13 "];
16[label="16, 16: 4 5 6 7 13 "];
17[label="17, 17: 4 5 6 7 "];
18[label="18, 18: 4 5 6 7 13 "];
19[label="19, 19: 4 5 6 7 13 "];
20[label="20, 20: 4 5 6 7 13 14 15 "];
21[label="21, 21: 4 5 6 7 14 15 16 17 18 19 "];
22[label="22, 22: 4 5 6 7 16 17 18 19 20 21 22 23 "];
23[label="23, 36: 20 21 22 23 "];
24[label="24, 37: 24 25 26 27 "];
25[label="25, 24: 24 25 26 27 "];
26[label="26, 26: "];
27[label="27, 27: "];
0->1 ;
1->2 ;
2->3 ;
3->4 ;
4->5 ;
5->6 ;
6->7 ;
7->8 ;
8->9 ;
8->10 ;
9->26 ;
10->11 ;
11->12 ;
12->13 ;
13->14 ;
14->15 ;
14->17 ;
15->16 ;
16->19 ;
17->18 ;
18->19 ;
19->20 ;
20->21 ;
21->22 ;
22->23 ;
23->24 ;
24->25 ;
25->26 ;
26->27 ;
}
