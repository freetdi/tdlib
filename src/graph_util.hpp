#pragma once

namespace treedec {

namespace util {

namespace impl {

template<class G, class M>
class edge_map {
public:
//	typedef typename boost::property_map<G, boost::vertex_index_t>::const_type VertexIndexMap;
	typedef typename boost::graph_traits<G>::vertex_descriptor vd;
//	typedef typename Map_type::value_type va<G>::vertex_descriptor vd;

	edge_map( G const& g, M const& m) : _m(m), _g(g) {}
//			boost::iterator_property_map < value_vertex_type *,
//			VertexIndexMap, value_vertex_type, value_vertex_type&> vd_map, const
//			G& _g): vd_map(vd_map), _g(_g) {}
 
	typedef typename M::value_type value_vertex_type;

	std::pair<value_vertex_type, value_vertex_type>
		operator()(typename boost::graph_traits<G>::edge_descriptor e) const {
			auto p=std::make_pair(boost::get(_m, boost::source(e, _g)),
					boost::get(_m, boost::target(e, _g)));
			trace2("edgmap", p.first, p.second);
			return p;
		}

private:
//	boost::iterator_property_map<value_vertex_type*,
//	                             VertexIndexMap,
//										  vd, value_vertex_type&> vd_map;
	M const& _m;
	G const& _g;
};

} // impl

template<class G, class M>
impl::edge_map<G, M> make_edge_map(M const& m, G const& g)
{
	return impl::edge_map<G, M>(g, m);
}

namespace detail{

// template<class S>
// struct visited_mask{
// static_assert(sizeof(S)==0);
// };

template<class S>
struct visited_mask{
	visited_mask(const visited_mask& o) : _s(o._s) {}
	visited_mask(S& s) : _s(s) {}
	bool operator()(unsigned x) const{
		assert(x<size());
		return _s[x];
	}
	bool operator[](unsigned x) const{
		return operator()(x);
	}
	void visit(unsigned x){
		assert(x<size());
		_s[x] = true;
	}
	size_t size() const{
		return _s.size();
	}
	S& _s;
};

}

detail::visited_mask<std::vector<BOOL> > make_incidence_mask(std::vector<BOOL>& v)
{
	return detail::visited_mask<std::vector<BOOL> >(v);
}

} //util

}
