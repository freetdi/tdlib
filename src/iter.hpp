// Felix Salfelder, 2014 - 2016
//
// (c) 2014-2016 Felix Salfelder
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

// graph related iteration

#ifndef TD_ITER_H
#define TD_ITER_H

#include <boost/graph/graph_traits.hpp>
#include <boost/graph/adjacency_list.hpp> // adjacent_vertices
#include <tr1/utility> // pair

template<class A>
void debug_count(A i, A e, unsigned s)
{
    unsigned c=0;
    for(; i!=e; ++i){
        ++c;
    }
    assert(c==s);
}


template<class T>
class subsets_iter{ //
public: // types
   typedef typename std::vector<T> scratch_type;
   typedef typename scratch_type::const_iterator base;
   class subset_iter : public std::vector<T>::const_iterator{ //
   public:
      subset_iter(typename std::vector<T>::const_iterator const & i)
          : std::vector<T>::const_iterator(i)
      {
      }
      subset_iter(typename std::vector<T> const & t)
          : std::vector<T>::const_iterator(t.begin())
      { untested();
      }
      subset_iter(T const & i)
      { untested();
          _t.resize(0);
          _t.push_back(i);
      }
   public: // ops
      typename T::value_type const& operator*()
      {
         return *std::vector<T>::const_iterator::operator*();
      }
      bool operator!=(const T& other)
      { untested();
         return !std::vector<T>::const_iterator::operator==(other);
      }
      bool operator==(const T& other)
      { untested();
         return std::vector<T>::const_iterator::operator==(other);
      }
   }; // subset_iter
public: // construct
   subsets_iter(const subsets_iter& p)
       : _tt(NULL),
         _t(p._t),
         _i(p._i), _e(p._e), _l(p._l), _u(p._u)
   {
   }
#if __cplusplus >= 201103L
   subsets_iter(const subsets_iter&& p)
       : _tt(p._tt),
         _t(p._t),
         _i(p._i), _e(p._e), _l(p._l), _u(p._u)
   {
       p._tt = NULL;
   }
#endif
   subsets_iter(T i, T e, size_t min=0, size_t max=-1, scratch_type* t=NULL)
       : _tt(NULL), _t(t?(*t):(*(new scratch_type))),
         _i(i), _e(e), _l(min), _u(max)
   {
#if __cplusplus >= 201103L
      if(t){
          t->resize(0);
          // use external scratch
          _tt = NULL;
      }else{
          // own scratch. record it.
          _tt = &_t;
      }
#endif
      assert(_l<=_u);
      fill();
   }
#if 0
   subsets_iter(T e)
       : _e(e), _l(0), _u(-1)
   { untested();
      _t.resize(1);
      _t[0] = e;
   }
#endif
   ~subsets_iter(){
      if(_tt){
          // own scratch. delete it.
          delete(&_t);
      }else{
      }
   }
public: // assign
   subsets_iter& operator=(const subsets_iter& other)
   { untested();
      _i = other._i;
      _e = other._e;
      _l = other._l;
      _u = other._u;
      _t = other._t;
//      _tt = other._tt;

//      other._tt=NULL;
      return *this;
   }
   subsets_iter& operator=(const T& other)
   { untested();
      base::operator=(other);
      _i = other._i;
      _e = other._e;
      _l = other._l;
      _u = other._u;
      _t = other._t;
      return *this;
   }
public: // ops
   bool operator==(const T& other)
   { itested();
      if(!_t.size()){ untested();
         // BUG?
         return false;
      }else if(_t.size()){
         return *_t.begin() == other;
      }else{ untested();
         return false;
      }
   }
   bool operator!=(const T& other)
   {
      return !operator==(other);
   }
   bool operator==(const subsets_iter& other)
   { untested();
      if(_t.size()){ untested();
          return _t.size() == other._t.size()
              && *_t[0]==*other._t[0];
      }else{ untested();
          return _i==other._i && _e==other._e;
      }
   }
   bool operator!=(const subsets_iter& other)
   { untested();
      return !operator==(other._i);
   }
   subsets_iter operator++()
   {
      if(_t.size()==0){ untested();
         _t.push_back(_i);
         assert(_i!=_e); // probably not.
                         // HACK: always push back something...
         if(_u==0){ untested();
            _t.back()=_e;
         }else{untested();
         }
      }else if(_t.size()<=_u){
         BOOST_AUTO(back,_t.back());
         ++back;
         if(back!=_e){
            if(_t.size()==_u){
               ++_t.back();
            }else{
               _t.push_back(back);
            }
         }else if(_t.back()==_e){ untested();
            unreachable();
         }else if(_t.size()==0){ untested();
            unreachable();
         }else if(_t.size()==1){
            ++_t.back();
         }else if(_t.size()==_l){
             BOOST_AUTO(back2, _t.back());
             carry(back2);
             if(_t.size()<_l){
                 _t[0]=_e;
             }else{
             }
         }else{
            _t.pop_back();
            BOOST_AUTO(back2, _t.back());
            ++back2;
            if(back2!=_e){
               ++_t.back();

            }else{ unreachable();
               // carry();
            }
         }
      }else if(_t.back() != _e){ untested();
         // assert(_t.size()==_u);
         incomplete();
      }else{ untested();
      }

//      fill();
      assert(_t.size()<=_u || _u==0);

      return *this;
   }
   void carry(T end)
   {
       BOOST_AUTO(b, _t.back());
       ++b;
       if(_t.size() == 1){
           ++_t.back();
       }else if(_t.back() == end){
           _t.pop_back();
           BOOST_AUTO(newend, _t.back());
           ++newend;
           if(newend==end){
               newend=_t.back();
           }else{
           }
           carry(newend);

           b = _t.back();
           ++b;
           if(_t.back()!=end){
               _t.push_back(b);
           }
       }else{
           assert(b==end);
           ++_t.back();
       }

   }
   void fill()
   {
      while(_t.size() < _l){
         if(!_t.size()){
            _t.push_back(_i);
         }else{
            BOOST_AUTO(back, _t.back());
            if(back==_e){ itested();
               break;
            }else{
               ++back;
               if(back==_e){
                       _t[0] = _e;

                   break;
               }
               _t.push_back(back);
            }
         }
      }

      assert(_t.size()>=_l || _t[0] == _e);
   }
   std::pair<subset_iter, subset_iter> operator*()
   {
      return std::make_pair(subset_iter(_t.begin()),
                            subset_iter(_t.end()));
   }
private: // detail
   void carry()
   { untested();
   }
private: // state
   mutable scratch_type* _tt;
   scratch_type& _t;
   T _i;
   /*const*/ T _e;
   /*const*/ size_t _l;
   /*const*/ size_t _u;
}; // subsets_iter

template<class A>
std::pair<subsets_iter<A>, subsets_iter<A> >
    make_subsets_iter(A a, A b, unsigned l, unsigned u,
          typename subsets_iter<A>::scratch_type* s=NULL)
{
   return std::make_pair(
       subsets_iter<A>(a,b,l,u,s),
       subsets_iter<A>(b,b));
}

#if 0
template<class G>
class nvsinconwsut_iter{
public: // types
   typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
public: // construct
public: // ops
private: // status
   std::vector<vertex_descriptor> _base;
};

template<class G>
std::pair<nvsinconwsut_iter<G>, nvsinconwsut_iter<G> >
    nonempty_vertex_sets_in_complement_of_neighborhood_with_size_up_to(*I, 2*k, g)
{
   return std::make_pair(nvsinconwsut_iter(*I, 2*k, g), nvsinconwsut_iter());
}
#endif

namespace detail{
// iterate neighbourhood of C, including C
// only works on ordered containers...
template<class A, class G>
class neighbourhood01_iter { //
public: // types
    typedef typename boost::graph_traits<G>::adjacency_iterator adjacency_iterator;
    typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
public: // construct
//     neighbourhood01_iter(const neighbourhood01_iter& p)
//     { untested();
//      incomplete. reuse _a?
//     }
    neighbourhood01_iter() : _v(NULL)
    { untested();
    }
    neighbourhood01_iter(A b, A e, unsigned size, const G& g, bool incb)
       : _b(b), _i(b), _e(e),
         _a(size), _g(g), _include_base(incb)
    {
        if(b==e){
            assert(size==0);
            return;
        }else{
        }

        bool found=false;

        if(_include_base){
            _v = *_i;
        }else{ untested();
            A ii(_b);
            for(; ii!=_e; ++ii){
                // just take one.... (inefficient)
                if(boost::adjacent_vertices(*ii, g).first
                 !=boost::adjacent_vertices(*ii, g).second){
                    _v = *boost::adjacent_vertices(*ii, g).first;
                    found = true;
                    break;
                }
            }
        }

        A ii(_b);

        unsigned n=0;
        for(; ii!=_e; ++ii){
            assert(!size || n<size);
            if(!size){ untested();
                _a.push_back(boost::adjacent_vertices(*ii, g).first);
            }else{
                _a[n] = boost::adjacent_vertices(*ii, g).first;
            }

            if(_a[n] == boost::adjacent_vertices(*ii, g).second){ untested();
            }else if(*(_a[n])<_v){
                _v = *_a[n];
                found = true;
            }else{
            }

            ++n;
        }
        if(_include_base){
        }else if(!found){ untested();
            _b = _e;
        }
    }

    ~neighbourhood01_iter()
    { itested();
    }
public: // ops
    bool operator!=(const neighbourhood01_iter& p) const
    {
        return _b != p._b;
    }
    bool operator==(const neighbourhood01_iter& p) const
    {
        return(_b == p._b);
    }
    neighbourhood01_iter& operator++()
    {
        if(_b==_e){ untested();
            return *this;
        }else{
        }

        vertex_descriptor previous = _v;
        bool found = false;
        if(_include_base){
            found = update(_i, _e, previous, _v);
        }else{ untested();
        }
        A ii(_b);

        unsigned n=0;
        for(; ii!=_e; ++ii){
            BOOST_AUTO(aend, boost::adjacent_vertices(*ii, _g).second);
            found |= update(_a[n], aend, previous, _v);
            ++n;
        }

        if(!found){
            _b = _e;
        }else{
            assert(_v>previous);
        }

        return *this;
    }
    const vertex_descriptor& operator*()
    {
        //std::cerr << "iter deref ++ " << _v << "\n";
        return _v;
    }
private: // impl
    template<class iter>
    bool update(iter& i, iter e, vertex_descriptor previous, vertex_descriptor& v)
    {
        if(i==e){
            return false;
        }else if(*i==previous){
            ++i;
            if(i==e){
                return false;
            }
        }else{
            //std::cerr << *i << " " << previous << "\n";
            assert(*i>previous);
        }

        if(v==previous){
            v = *i;
        }else if(*i<v){
            assert(*i>previous);
            v = *i;
        }else{
        }
        return true;
    }
private: // data
    A _b;
    A _i;
    A _e;
    std::vector<adjacency_iterator> _a;
    vertex_descriptor _v;
    G const& _g;
    bool _include_base;
};
} // detail



template<class A, class G>
std::pair<detail::neighbourhood01_iter<A, G>,
          detail::neighbourhood01_iter<A, G> >
inline make_neighbourhood01_iter(A b, A e, G const& g, unsigned size=0,
        bool include_base=true)
{
    typedef detail::neighbourhood01_iter<A, G> nIter;
    return std::make_pair(
            nIter(b,e,size,g,include_base),
            nIter(e,e,0,g,include_base) );
}

template<class A, class G>
std::pair<detail::neighbourhood01_iter<A, G>,
          detail::neighbourhood01_iter<A, G> >
inline make_neighbourhood_iter(A b, A e, G const& g, unsigned size=0)
{
    return make_neighbourhood01_iter(b,e,g,size,false);
}

#endif

// vim:ts=8:sw=4:et
