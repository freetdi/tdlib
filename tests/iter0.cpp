#include <set>
#include <iostream>
#include <stdlib.h> // malloc
#include <string.h> // memcpy
#include <vector>
#include <assert.h>
#include <tr1/utility>

#include <boost/typeof/typeof.hpp>

#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/edge_list.hpp>
#include <iostream>

#include <treedec/iter.hpp>
#include <treedec/misc.hpp>
#include <treedec/graph_util.hpp>


int main()
{
	boost::adjacency_list<boost::setS, boost::vecS, boost::undirectedS> g;

	auto v1 = boost::add_vertex(g);
	auto v2 = boost::add_vertex(g);
	auto v3 = boost::add_vertex(g);
	auto v4 = boost::add_vertex(g);

	std::cout << v1 << ": " <<boost::get(boost::vertex_index, g, v1) << "\n";
	std::cout << v2 << ": " <<boost::get(boost::vertex_index, g, v2) << "\n";
	std::cout << v3 << ": " <<boost::get(boost::vertex_index, g, v3) << "\n";

	{
		auto R = make_components_range(g);
		auto& cmpi = R.first;
		auto ii = (*cmpi).first;
		std::cout << *ii << "\n";
		std::cout << "\n";
	}
	std::cout << "========\n";

	boost::add_edge(v1,v2,g);
	{
		std::cout << "\n";

		auto vir=boost::vertices(g);
		std::vector<BOOL> visited(boost::num_vertices(g), false);

		auto M = treedec::util::make_incidence_mask(visited);

		auto R = make_components_range(vir.first, vir.second, g, M);
		auto& cmpi = R.first;

		unsigned k=0;
		for(; cmpi!=R.second;){
			auto P=(*cmpi);
			auto ii=P.first;
			auto ee=P.second;
			for(; ii!=ee; ++ii){
				assert(k==*ii);
				++k;
			}

			++cmpi;
		}
		std::cout << k << "\n";
		std::cout << "========\n";

		{
			auto R = make_components_range(g);
			auto& cmpi = R.first;

			unsigned k=0;
			for(; cmpi!=R.second;){
				auto P = (*cmpi);
				auto ii = P.first;
				auto ee = P.second;
				for(; ii != ee; ++ii){
					assert(k==*ii);
					++k;
				}

				++cmpi;
			}
		}
		std::cout << "========\n";

	}
	{
		boost::add_edge(v3,v4,g);
		// 0 -- 1
		// 2 -- 3

		assert(boost::num_edges(g) == 2);
		assert(boost::num_vertices(g) == 4);
		auto vi=boost::vertices(g).first;
		auto ve=boost::vertices(g).second;
		++vi;
		++vi;
		++vi;
		std::vector<BOOL> visited(boost::num_vertices(g), false);
		auto vm = treedec::util::make_incidence_mask(visited);

		auto R = make_components_range(vi, ve, g, vm);
		auto& cmpi = R.first;
		int k = 0;

		for(; cmpi!=R.second;){
			auto P = (*cmpi);
			auto ii = P.first;
			auto ee = P.second;
			for(; ii != ee; ++ii){
				assert(int(*ii)==3-k);
				++k;
			}

			++cmpi;
		}
		assert(k == 2);
	}
	std::cout << "========\n";
	{
		auto vi = boost::vertices(g).first;
		auto ve = boost::vertices(g).first;
		++vi;
		++ve;
		++ve;
		++ve;
		std::vector<BOOL> visited(boost::num_vertices(g), false);
		auto vm = treedec::util::make_incidence_mask(visited);

		auto R = make_components_range(vi,ve,g,vm);
		//	auto R=make_components_range(g);
		auto& cmpi = R.first;

		for(; cmpi!=R.second;){
			auto P = (*cmpi);
			auto ii = P.first;
			auto ee = P.second;
			for(; ii != ee; ++ii){
				std::cout << " " << *ii;
			}
			std::cout << "\n";

			++cmpi;
		}
	}

	std::cout << "========\n";
	{
		auto v5 = boost::add_vertex(g);
		auto v6 = boost::add_vertex(g);
		auto v7 = boost::add_vertex(g);
		auto v8 = boost::add_vertex(g);

		auto vi=boost::vertices(g).first;
		auto ve=boost::vertices(g).first;
		++vi;
		++ve;
		++ve;
		++ve;
		std::vector<BOOL> visited(boost::num_vertices(g), false);
		visited[0]=true;
		visited[1]=true;
		boost::add_edge(v1,v5,g);
		boost::add_edge(v2,v6,g);
		boost::add_edge(v3,v6,g);
		boost::add_edge(v3,v7,g);
		boost::add_edge(v3,v8,g);

		auto vm=treedec::util::make_incidence_mask(visited);
		auto R=make_components_range(vi,ve,g,vm);
		//	auto R=make_components_range(g);
		auto& cmpi = R.first;

		for(; cmpi!=R.second;){
			auto P=(*cmpi);
			auto ii=P.first;
			auto ee=P.second;
			for(; ii!=ee; ++ii){
				std::cout << " " << *ii;
			}
			std::cout << "\n";

			++cmpi;
		}
	}
	std::cout << "========\n";
	{
		g.clear();
		auto v0=boost::add_vertex(g);
		(void) v0;
		auto v1 = boost::add_vertex(g);
		auto v2 = boost::add_vertex(g);
		auto v3 = boost::add_vertex(g);
		auto v4 = boost::add_vertex(g);
		auto v5 = boost::add_vertex(g);
		auto v6 = boost::add_vertex(g);
		auto v7 = boost::add_vertex(g);
		auto v8 = boost::add_vertex(g);
		(void) v8;

		auto vi=boost::vertices(g).first;
		auto ve=boost::vertices(g).first;
		++vi;
		++ve;
		++ve;
		++ve;
		std::vector<BOOL> visited(boost::num_vertices(g), false);
		boost::add_edge(v1,v3,g);
		boost::add_edge(v2,v4,g);
		boost::add_edge(v3,v5,g);
		boost::add_edge(v4,v6,g);
		boost::add_edge(v5,v7,g);

		auto vm = treedec::util::make_incidence_mask(visited);
		auto R = make_components_range(vi,ve,g,vm);
		auto& cmpi = R.first;

		unsigned c = 0;
		for(; cmpi!=R.second; ++cmpi){
			auto P = (*cmpi);
			auto ii = P.first;
			auto ee = P.second;
			unsigned m = 0;
			for(; ii!=ee; ++ii){
				assert (2*m+c+1 == *ii);
				++m;
			}

			++c;
		}
	}
	std::cout << "========\n";
	{
		g.clear();
		auto v0 = boost::add_vertex(g);
		(void) v0;
		auto v1 = boost::add_vertex(g);
		auto v2 = boost::add_vertex(g);
		auto v3 = boost::add_vertex(g);
		auto v4 = boost::add_vertex(g);
		auto v5 = boost::add_vertex(g);
		auto v6 = boost::add_vertex(g);
		auto v7 = boost::add_vertex(g);
		auto v8 = boost::add_vertex(g);
		(void) v8;

		auto vi = boost::vertices(g).first;
		auto ve = boost::vertices(g).first;
		++vi;
		++ve;
		++ve;
		++ve;
		std::vector<BOOL> visited(boost::num_vertices(g), false);

		// 0  1 -- 3 -- 5 -- 7
		//    2 -- 4 -- 6
		//
		// [vi, ve) = {1,2}
		//
		boost::add_edge(v1,v3,g);
		boost::add_edge(v2,v4,g);
		boost::add_edge(v3,v5,g);
		boost::add_edge(v4,v6,g);
		boost::add_edge(v5,v7,g);
		visited[0]=true;
		visited[2]=true;
		visited[7]=true;
		// visit 1 3 5...


		auto vm = treedec::util::make_incidence_mask(visited);
		auto R = make_components_range(vi, ve, g, vm);
		auto& cmpi = R.first;

		unsigned k = 0;
		for(; cmpi!=R.second;){
			unsigned kk = 0;
			auto P = (*cmpi);
			auto ii = P.first;
			auto ee = P.second;
			for(; ii!=ee; ++ii){
				assert(k==0);
				trace2("visit", *ii, k);
				assert(*ii == 2* kk + 1);
				++kk;
			}
			std::cout << "\n";

			++k;
			++cmpi;
		}
	}
}
