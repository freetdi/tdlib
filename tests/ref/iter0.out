0: 0
1: 1
2: 2
0

========

4
========
========
========
 1 0
 2 3
========
 2 3 5 6 7
========
========

